controllers/machine_controller_phases.go

```diff
@@ -97,9 +96,18 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
 	obj, err := external.Get(ctx, r.Client, ref, m.Namespace)
 	if err != nil {
 		if apierrors.IsNotFound(errors.Cause(err)) {
-			return external.ReconcileOutput{}, errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: externalReadyWait},
-				"could not find %v %q for Machine %q in namespace %q, requeuing",
-				ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+			// if the reference missing is the infrastructureRef and the machine was up and running, this is an error.
+			if *ref == m.Spec.InfrastructureRef && m.Status.InfrastructureReady {
+				err := errors.Errorf("could not find %v %q for Machine %q in namespace %q, requeuing", ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+				log.Error(err, "Machine infrastructure reference has been deleted after being ready, setting failure state")
+				m.Status.FailureReason = capierrors.MachineStatusErrorPtr(capierrors.InvalidConfigurationMachineError)
+				m.Status.FailureMessage = pointer.StringPtr(fmt.Sprintf("Machine infrastructure resource %v with name %q has been deleted after being ready",
+					m.Spec.InfrastructureRef.GroupVersionKind(), m.Spec.InfrastructureRef.Name))
+				return external.ReconcileOutput{}, err
+			}
```

Comment:
Why are we changing logic?

Reply:
We are not changing the logic, we are only moving here a bit of logic from reconcileInfrastructure.

This logic handles a very specific error case, which is considered terminal and triggers a different response that included setting failure reason and message. 
NB. this should be reworked again soon we will move this kind o message to conditions.

Reply:
So should we be setting `m.Status.InfrastructureReady` to `false`? Since we are setting the Machine's failure reason and message.

Also we should probably change the `log.Error` and `FailureMessage` to not specifically refer to `Machine Infrastructure resource` since this method is being called by `reconcileBootstrap` as well as `reconcileInfrastructure`.

Reply:
> So should we be setting m.Status.InfrastructureReady to false

No, by setting FailureReason/FailureMessage we are signaling a terminal failure
Also, this would be a behavior change.

> Also we should probably change the log.Error and FailureMessage to not specifically refer to Machine Infrastructure resource

This bit of code is specific for the infrastructure ref only, so the message could be specific